pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,383 @@
+}
+
+func (gc *GarbageCollector) deleteObject(item *node) error {
+	fqKind := unversioned.FromAPIVersionAndKind(item.identity.APIVersion, item.identity.Kind)
+	client, err := gc.clientPool.ClientForGroupVersion(fqKind.GroupVersion())
+	resource, err := gc.apiResource(item.identity.APIVersion, item.identity.Kind, item.identity.Namespace)
+	if err != nil {
+		return err
+	}
+	uid := item.identity.UID
+	preconditions := v1.Preconditions{UID: &uid}
+	deleteOptions := v1.DeleteOptions{Preconditions: &preconditions}
+	return client.Resource(resource, item.identity.Namespace).Delete(item.identity.Name, &deleteOptions)
+}
+
+func (gc *GarbageCollector) getObject(item *node) (*runtime.Unstructured, error) {
+	fqKind := unversioned.FromAPIVersionAndKind(item.identity.APIVersion, item.identity.Kind)
+	client, err := gc.clientPool.ClientForGroupVersion(fqKind.GroupVersion())
+	resource, err := gc.apiResource(item.identity.APIVersion, item.identity.Kind, item.identity.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	return client.Resource(resource, item.identity.Namespace).Get(item.identity.Name)
+}
+
+func (gc *GarbageCollector) processItem(item *node) error {
+	// Get the latest item from the API server
+	latest, err := gc.getObject(item)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			glog.V(6).Infof("item %v not found, ignore it", item.identity)
+			return nil
+		}
+		return err
+	}
+	if latest.GetUID() != item.identity.UID {
+		glog.V(6).Infof("UID doesn't match, item %v not found, ignore it", item.identity)
+		return nil
+	}
+	ownerReferences := latest.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		glog.V(6).Infof("object %s's doesn't have an owner, continue on next item", item.identity)
+		return nil
+	}
+	for _, reference := range ownerReferences {
+		// TODO: need to verify the reference resource is supported by the system.
+		fqKind := unversioned.FromAPIVersionAndKind(reference.APIVersion, reference.Kind)
+		client, err := gc.clientPool.ClientForGroupVersion(fqKind.GroupVersion())
+		if err != nil {
+			return err
+		}
+		resource, err := gc.apiResource(reference.APIVersion, reference.Kind, item.identity.Namespace)
+		if err != nil {
+			return err
+		}
+		owner, err := client.Resource(resource, item.identity.Namespace).Get(reference.Name)
+		// TODO: need to compare the UID.
+		if err == nil {
+			if owner.GetUID() != reference.UID {
+				glog.V(6).Infof("object %s's owner %s/%s, %s is not found", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+				continue
+			}
+			glog.V(6).Infof("object %s has at least an existing owner, will not garbage collect", item.identity.UID)
+			return nil
+		} else if errors.IsNotFound(err) {
+			glog.V(6).Infof("object %s's owner %s/%s, %s is not found", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+		} else {
+			return err
+		}
+	}
+	glog.V(6).Infof("none of object %s's owners exist any more, will garbage collect it", item.identity.UID)
```

Comment:
maybe V(2)-- this seems somewhat useful. Also log the namespace & name for easy log reading for humans.